business/authentication/session_persistor.go

```diff
@@ -0,0 +1,335 @@
+package authentication
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/kiali/kiali/business"
+	"github.com/kiali/kiali/config"
+	"github.com/kiali/kiali/log"
+	"github.com/kiali/kiali/util"
+)
+
+type SessionPersistor interface {
+	CreateSession(r *http.Request, w http.ResponseWriter, strategy string, expiresOn time.Time, payload interface{}) error
+	ReadSession(r *http.Request, w http.ResponseWriter, payload interface{}) (sData *sessionData, err error)
+	TerminateSession(r *http.Request, w http.ResponseWriter)
+}
+
+type CookieSessionPersistor struct{}
+
+type sessionData struct {
+	Strategy  string    `json:"strategy"`
+	ExpiresOn time.Time `json:"expiresOn"`
+	Payload   string    `json:"payload,omitempty"`
+}
+
+// CreateSession starts a user session using HTTP Cookies for persistance across HTTP requests.
+// For improved security, the data of the session is encrypted using the AES-GCM algorithm and
+// the encrypted data is what is sent in cookies. The strategy, expiresOn and payload arguments
+// are all required.
+func (p CookieSessionPersistor) CreateSession(_ *http.Request, w http.ResponseWriter, strategy string, expiresOn time.Time, payload interface{}) error {
```

Comment:
I don't see where the cookie data is persisted on the backend in this method? I was expecting `CreateSession` to persist the cookie data somehow on the backend (e.g. in memory or on the filesystem) and then call `http.SetCookie(sessionID)` returning just the session id to the client. Instead all of the session data is being persisted client side by returning the full session data in the `SetCookie(chunksCookie)` call.

Perhaps we can get around needing to chunk the cookie data if the backend just returns the session id instead of the full b64 encoded/encrypted cookie contents and saves the encrypted cookie contents on the backend?

If we want to bring in another external dependency, there might be some libraries that handle saving/reading on the backend either in memory or to disk such as: https://pkg.go.dev/github.com/gorilla/sessions.

Reply:
Sending only _SIDs_ is my preference. However, as you say, back-end storage is needed.

Both memory and pod filesystem storage are volatile in Kubernetes. A pod restart would kill all sessions... and it can be benign: K8s may be trying to optimize resource usage by relocating the Kiali pod to another node.

So, we need something capable of surviving a pod restart and I can only think on out-of-pod storage. Most folks on the team are against adding storage because of the perception that storage in k8s is hard to manage. I don't have the needed experience to reject that but if you can convince the team that storage isn't bad, let's start yet another round of debate ;-)

Reply:
BTW, that gorilla/sessions library looks good. However, I looked at the code, and it lacks support splitting large payloads into multiple cookies :-(

Reply:
> Both memory and pod filesystem storage are volatile in Kubernetes. A pod restart would kill all sessions... and it can be benign: K8s may be trying to optimize resource usage by relocating the Kiali pod to another node.

Yes if the pod gets rescheduled onto a different node it'd wipe the ephemeral storage. That's probably not a great UX to lose your session whenever that happens. On the other hand, part of the tradeoff of storing sessions in the browser is that they cannot be invalidated on the backend.

> Most folks on the team are against adding storage because of the perception that storage in k8s is hard to manage.

I am one of them :). I wonder if it'd be feasible to store session data in a configmap like:
```
data:
  "sid-1": {}
  "sid-2": {}
```
Since Kiali already talks to the kube API, the integration would be straightforward. There's apparently a 1mb limit on the amount of data you can store in a configmap though. Either way, changing how Kiali stores sessions is probably out of scope for this PR.